### DISCUSSION TITLE
Two frequent "system CPU load between ticks" calls result in inconsistent values

### DISCUSSION THREAD

**Discussion by facewindu** [oshi/oshi#2619] [Q&A] [answered]
I'm doing this sort of things
```

ScheduledExecutorService scheduler = NamedExecutors.newSingleThreadScheduledExecutor("foo");
SystemInfo si = new SystemInfo();
HardwareAbstractionLayer hal = si.getHardware();
CentralProcessor processor = hal.getProcessor();
long[] oldTicks = null;
scheduler.scheduleAtFixedRate(() -> captureData(processor, oldTicks), 0, 100, TimeUnit.MILLISECONDS);

private static void captureData(CentralProcessor processor, long[] oldTicks) {
  double load = processor.getSystemCpuLoadBetweenTicks(oldTicks);
  oldTicks = processor.getSystemCpuLoadTicks();
  // do sth with the captured load
}
```

So every 100ms, I'm getting the system CPU load.
I consistently get some legit values for some samples, and then some zeroes for a few samples, then back to some legit values.

If I increase the sampling period to sth bigger, like every second instead of every 100ms, all samples are legit (non 0).

Am I doing sth wrong? Or is there a reason why sampling the system CPU load too often is known to cause this kind of problems ?

**Comment by mprins** [comment]
You have failed to provide your OS, 
but please see the note about rounding ticks to 1/64 on Windows in eg.: https://github.com/oshi/oshi/issues/1870#issuecomment-1021387093

**Comment by facewindu** [comment]
Thanks for the reply. I'm on latest macOS Sonoma 14.4.1 with oshi `6.4.13`

**Comment by dbwiddis** [comment]
CPU ticks are cached/memoized.  I think the Javadoc says something about not being fetched more often than about 1 second intervals.

**Comment by dbwiddis** [comment]
You can change the memorizer expiration in properties.  

https://github.com/oshi/oshi/blob/276f4c4ab88115e91e32e12f41d03d5354579ffd/oshi-core/src/main/java/oshi/hardware/common/AbstractCentralProcessor.java#L49

**Comment by dbwiddis** [comment]
https://github.com/oshi/oshi/blob/276f4c4ab88115e91e32e12f41d03d5354579ffd/oshi-core/src/main/java/oshi/util/Memoizer.java#L26

**Comment by dbwiddis** [comment, marked as answer]
I was answering from my phone earlier, thus the brief permalinks to the relative bits.  But the answer to your question is that by default, ticks are only going to update every 300 milliseconds so at a refresh value more frequent than this, your "between ticks" value will be zero.

OSHI doesn't guarantee re-reading the config, but this is one place where it does; reading this config parameter every minute.  You can call `GlobalConfig.set(GlobalConfig.OSHI_UTIL_MEMOIZER_EXPIRATION, 90);` (or any value of your choice less than 100 milliseconds) to cache for a shorter amount of time and get the results you expect.

**Reply by facewindu** [reply]
Thank you for the detailed answer. I'll experiment with that.
